Add file-like pager: `click.get_pager_file()`

craigds opened the pull request:
In koordinates/kart we're in need of an output pager, and are already using click.

Trouble is, passing a generator to `echo_via_pager` requires that you're producing output in _one_ place. That's not the case with a reasonably complex command - functions and modules delegating to each other all over the place.

To use `echo_via_pager`, I essentially had to use some kind of concurrency - async or threads - and have my code feed output to a worker thread which would yield the output from a generator function, which `echo_via_pager` would consume.

That _worked_, but then I realised click could just expose the file-like object it's using internally (which is usually a pipe to `less`).

This PR does that. `click.get_pager_file` exposes a writable file like object which feeds into a pager. The semantics of `click.echo_via_pager` haven't changed, but it now uses `click.get_pager_file` under the hood and merely feeds the generator contents into the file.

davidism commented:
Thanks for looking at this, it's definitely an interesting approach. Have you confirmed that this works on Windows? Have you tested it with different pagers on Linux?

craigds commented:
Thanks for taking a look.

I've tested both `_pipepager` and `_tempfilepager` paths, but only with `less` and only on MacOS. I'll try with linux builds and with other pagers, and probably I can rope @rcoup into testing on Windows for me...

craigds committed:
Refactor `echo_via_pager` to expose file objects

This better exposes the file-like objects yielded by the three private
pager functions, and also slightly reduces some code duplication

craigds committed:
Add file-like pager: `click.get_pager_file()`

craigds commented:
Extra testing is still TODO, but the new commits are a bit more reviewable, esp if you read them one at a time 👍

craigds commented:
On MacOS and Linux (I did it with the `python:3.8` docker image) output works as expected with `PAGER=cat`, `PAGER=less`, `PAGER=more`.

stefreak commented:
I love this one. What is missing except for a round of conflict resolution? Does it really need extra test or is it covered by test of echo_via_pager?

craigds commented:
It needs testing in windows. The little testing I did suggested that it doesn't really work on windows  yet. I got stuck trying to set up any kind of dev environment on windows to improve it. If anyone who has a windows setup already could help, that'd be much appreciated

craigds commented:
the mypy stuff is also blowing my mind a bit to be honest. I just spent about an hour trying to resolve all the type warnings and haven't yet managed to figure out what to do with things like variables that can hold either a text stream or a bytes stream

stefreak commented:
@craigds would you mind if I give this another shot once #2775 has been merged? I rented a Hetzner windows server for development tasks like this and could look into adding tests for windows.

craigds commented:
@stefreak yep that'd be awesome. I'll have a go at updating this PR for you

craigds commented:
@stefreak I've updated this PR to latest main and it works okay for me on MacOS:


<img width="962" alt="Screenshot 2024-09-09 at 11 36 07 AM" src="https://github.com/user-attachments/assets/5e1ea1ab-dcee-4343-b74f-daa0d98f3c4a">

AndreasBackx commented:
@stefreak @craigds this might be able make it into 8.2.0. If you rebase this PR on https://github.com/pallets/click/pull/2775, I'm happy to accept a "stacked PR" like that. It should be possible to get https://github.com/pallets/click/pull/2775 in 8.2.0, see https://github.com/pallets/click/issues/2789 for more info. I'm working through all 8.1.8 and 8.2.0 potential PRs right now and updating that issue and a new one as well.

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

craigds committed:
Fix confusing bug in _tempfilepager on windows

craigds committed:
Fix type hints for get_pager_file components

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

craigds committed:
pager: Fix ansi chars not being stripped correctly

MaybeStripAnsi wasn't correctly applied to text streams, meaning that
when called with `LESS=-c PAGER=less` (ie LESS is defined but doesn't
include `-R`) the ansi chars should be stripped but weren't.

This fixes the problem by ensuring text streams get unwrapped and
re-wrapped in a MaybeStripAnsi object (and also get flushed correctly
when handling KeyboardInterrupt)

craigds commented:
any chance of this getting merged some time soon? or any other review comments?

kdeldycke commented:
@davidism @Rowlando13 : I propose to target this for Click 8.4.0. The 8.3.2 release had no strong feedback, so we can continue polish things in 8.3.3, and start back with 8.4.0 to integrate heavier code structure changes.

kdeldycke commented:
The reason I think this one is doable is we now have randomized and parallel tests (https://github.com/pallets/click/pull/3151) and a collection of stress tests (https://github.com/pallets/click/pull/3139). So we have the infrastructure in place to push these stream-manipulating code path further and increase our chances to catch file pointers and buffer leaks. Speaking of which, @craigds, can you come up with a couple of tests in this PR? 😁

Rowlando13 commented:
@kdeldycke Can you fix merge conflicts and I will merge this for 8.4.0. Did not see this and accidentally merged something with a conflict.

AndreasBackx committed:
Merge branch 'main' into filelike-pager

Rowlando13 commented:
@AndreasBackx Thanks for getting this in! Would you mind moving the changelog entry to the correct section?

AndreasBackx commented:
My bad, thanks for spotting! fed9049f7a07550d560a91b30c5b0b3e17d54981

kdeldycke commented:
> Speaking of which, @craigds, can you come up with a couple of tests in this PR? 😁

This should have not been merged as-is: there is 0 test demonstrating that it is working as intended.

AndreasBackx commented:
> > Speaking of which, @craigds, can you come up with a couple of tests in this PR? 😁
> 
> This should have not been merged as-is: there is 0 test demonstrating that it is working as intended.

My apologies. I'll work on some new tests myself and put them up hopefully this evening in a new PR. Shouldn't be too difficult. Let me know if there's anything in particular you already had in mind.

kdeldycke commented:
> My apologies. I'll work on some new tests myself and put them up hopefully this evening in a new PR. Shouldn't be too difficult. Let me know if there's anything in particular you already had in mind.

I'd like to insist on testing because that's the lack of testing that made regressions commonplace for years.

In these situation what I do is start by re-coding the case presented in the original issue or PR's body into the test suit of Click in the appropriate place, then try to add a `@parametrize` test that makes me focus on the data. So I get a clean entry-point for possible usages and a higher level view. Then add all sorts of edge-cases or stupid input to that `@parametrize` decorator like: empty strings, empty lists, `UNSET` sentinels, fancy types, Enums, etc. Most of the time I stumble upon overlooked issues worth fixing.

AndreasBackx commented:
See #3405 @kdeldycke. I'm not entirely sure what edge cases to test for however:

> Then add all sorts of edge-cases or stupid input to that @parametrize decorator like: empty strings, empty lists, UNSET sentinels, fancy types, Enums, etc. Most of the time I stumble upon overlooked issues worth fixing.

Most of this will be converted to `str` so we're not really testing the paging behaviour here from what I can tell, hence I've not included some of the proposals you mentioned. But let me know whether or not I am missing something.

kdeldycke commented:
FYI, this PR introduced 2 regressions on Windows:
- https://github.com/pallets/click/pull/3739
- https://github.com/pallets/click/pull/3764

Can you test them?

AndreasBackx commented:
@kdeldycke Apologies as I was on holiday when you posted. Is there still anything to test now that they've been merged or anything in particular you want me to test?

kdeldycke commented:
> @kdeldycke Apologies as I was on holiday when you posted. Is there still anything to test now that they've been merged or anything in particular you want me to test?

Nothing particular, just wanted to make sure people knowing that part of the codebase could have a chance to review or double check the recent changes. Thanks for chiming in! :)
